### PR TITLE
Rover: improved handling of reverse

### PR DESCRIPTION
**Pull request by rmackay9**
This PR improves Rover's handling of reversing including fixing this issue (https://github.com/ArduPilot/ardupilot/issues/8311):

- AP_MotorsUGV always interprets a positive steering input as a request to turn the vehicle clockwise regardless of the throttle input.  Essentially this change involves making Ackermann steering vehicles like skid-steering vehicles instead of the other way around.  This simplifies higher level steering controllers because it reduces the number of places we need to check if the vehicle is moving forwards or backwards.
- In manual/semi-manual modes (manual, acro, steering) when interpreting the driver's steering input (which can depend upon the throttle input) we use a throttle-slew or acceleration-limited throttle.  This resolves an issue in which the vehicle might temporarily turn in the wrong direction as the vehicle transitioned between forward and reverse.
- speed scaling of steering for Ackermann steer vehicles is done in the motors library instead of the attitude controller.  This has the following benefits:
   - architecturally puts the scaling very close to the similar scaling for vectored-thrust vehicles
   - allows support for vehicles with both skid-steering and Ackermann style steering (like a ski-doo).
   - essentially scales down the I-term contribution to steering as speed increases.  previously when done in the attitude controller, the scaling was done on the error term.  This led to I-term built up at low speeds taking longer to "bleed off" at high speeds.
- the potential downside of the above change is that at very high speeds the steering control may become so limited that it cannot counteract a physical misalignment in the steering linkage.  For example at 20m/s (70km/h) only 5% of the total steering range is available.  A potential solution to this though is to allow steering inputs > 1.
- a relatively minor change included in this PR is that we perform no steering scaling in manual mode.  this change only affects the relatively new vector-thrust feature that has not yet been released.

### CODE REVIEW

File: APMrover2/mode.cpp [hunk 1]
```diff
@@ -90,6 +90,40 @@ void Mode::get_pilot_desired_steering_and_throttle(float &steering_out, float &t
...
+void Mode::get_pilot_desired_steering_and_throttle(float &steering_out, float &throttle_out)
+{
+    // do basic conversion
+    get_pilot_input(steering_out, throttle_out);
+
+    // check for special case of input and output throttle being in opposite directions
+    float throttle_out_limited = g2.motors.get_slew_limited_throttle(throttle_out, rover.G_Dt);
+    if ((is_negative(throttle_out) != is_negative(throttle_out_limited)) &&
+        ((g.pilot_steer_type == PILOT_STEER_TYPE_DEFAULT) ||
+         (g.pilot_steer_type == PILOT_STEER_TYPE_DEFAULT))) {
```
Comment (OXINARF) [inline APMrover2/mode.cpp]: Something is wrong here, right? The second one was supposed to be PILOT_STEER_TYPE_DIR_REVERSED_WHEN_REVERSING?

File: APMrover2/mode.cpp [hunk 2]
```diff
@@ -90,6 +90,40 @@ void Mode::get_pilot_desired_steering_and_throttle(float &steering_out, float &t
     }
 }
 
+// decode pilot steering held in channel_steer, channel_throttle and return in steer_out and throttle_out arguments
+// steering_out is in the range -4500 ~ +4500 with positive numbers meaning rotate clockwise
+// throttle_out is in the range -100 ~ +100
```
Comment (OXINARF) [inline APMrover2/mode.cpp]: Nitpicking: making the comment from header be the same here would be better.

File: APMrover2/mode.cpp [hunk 3]
```diff
@@ -90,6 +90,40 @@ void Mode::get_pilot_desired_steering_and_throttle(float &steering_out, float &t
...
+void Mode::get_pilot_desired_steering_and_speed(float &steering_out, float &speed_out)
+{
+    float desired_throttle;
+    get_pilot_input(steering_out, desired_throttle);
+    speed_out = desired_throttle * 0.01f * calc_speed_max(g.speed_cruise, g.throttle_cruise * 0.01f);
+    // check for special case of input and output throttle being in opposite directions
+    float speed_out_limited = g2.attitude_control.get_desired_speed_accel_limited(speed_out);
+    if ((is_negative(speed_out) != is_negative(speed_out_limited)) &&
+        ((g.pilot_steer_type == PILOT_STEER_TYPE_DEFAULT) ||
+         (g.pilot_steer_type == PILOT_STEER_TYPE_DEFAULT))) {
```
Comment (OXINARF) [inline APMrover2/mode.cpp]: Same.

File: APMrover2/mode_steering.cpp [hunk 4]
```diff
@@ -15,38 +15,41 @@ void ModeSteering::update()
     float desired_steering, desired_speed;
     get_pilot_desired_steering_and_speed(desired_steering, desired_speed);
 
+    bool reversed = is_negative(desired_steering);
```
Comment (OXINARF) [inline APMrover2/mode_steering.cpp]: Is this really supposed to be `desired_steering` and not `desired_speed`?

**Comment by rmackay9** [comment]
@OXINARF, thanks for the review!  As per usual you've found a couple of mistakes I've made, having your review really gives me confidence.  Txs and fixes coming shortly.

File: APMrover2/mode_steering.cpp (hunk 4 above)
Comment (rmackay9) [reply APMrover2/mode_steering.cpp]: That was definitely a bug, fixed in follow up commit.

File: APMrover2/mode.cpp (hunk 3 above)
Comment (rmackay9) [reply APMrover2/mode.cpp]: yes, a bug, thanks again.  fixed in a follow up commit

File: APMrover2/AP_MotorsUGV.cpp [hunk 5]
```diff
@@ -524,16 +537,12 @@ void AP_MotorsUGV::output_throttle(SRV_Channel::Aux_servo_function_t function, f
...
-            _throttle = _throttle_prev + throttle_change_max;
-            limit.throttle_upper = true;
-        } else if (_throttle < _throttle_prev - throttle_change_max) {
-            _throttle = _throttle_prev - throttle_change_max;
-            limit.throttle_lower = true;
-        }
+    const float throttle_orig = _throttle;
+    _throttle = get_slew_limited_throttle(_throttle, dt);
+    if (_throttle < throttle_orig) {
+        limit.throttle_upper = true;
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply APMrover2/AP_MotorsUGV.cpp]: OK, reversed the conditional command to make it easier to understand, thanks!

File: APMrover2/AP_MotorsUGV.cpp [hunk 6]
```diff
@@ -151,9 +152,12 @@ void AP_MotorsUGV::setup_servo_output()
 }
 
 // set steering as a value from -4500 to +4500
-void AP_MotorsUGV::set_steering(float steering)
+//   apply_scaling should be set to false for manual modes where
+//   no scaling by speed or angle should e performed
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply APMrover2/AP_MotorsUGV.cpp]: cool, fixed the spelling mistake, txs

File: APMrover2/AP_MotorsUGV.h [hunk 7]
```diff
@@ -115,4 +119,8 @@ class AP_MotorsUGV {
     float   _steering;  // requested steering as a value from -4500 to +4500
     float   _throttle;  // requested throttle as a value from -100 to 100
     float   _throttle_prev; // throttle input from previous iteration
+
+    // steering scaling related variables
+    bool    _scale_steering = true; // true if we should scale steering by speed or angle
+    float   _ground_speed;  // speed of vehicle over surface.  used to scale back steering at high speed or when reversing
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply APMrover2/AP_MotorsUGV.h]: yup, removed the ground_speed member and we just pass it in each time.

**Comment by khancyr** [comment]
Humm, it looks nice, but I will need a little more time to review correctly. 
One point that bother me is that steering is diverging from the Auto controler when then should be close to allow debugging the auto controler.
I will try to put more test on rover to check the controller in autotest

**Comment by rmackay9** [comment]
I've updated this PR with the following changes:

- rebased on master to resolve conflicts with the omni-rover PR
- fixed a bug in the "Rover: fix calc-steering-to-waypoint to use real heading without reverse" that was causing missions to fail for skid-steering vehicles.

**Comment by rmackay9** [comment]
I tested this some more yesterday on real vehicles and it seems OK so merging!

**Pull request merged**